cpp/src/arrow/status.cc

```diff
@@ -26,18 +28,19 @@ Status::Status(StatusCode code, const std::string& msg)
 
 Status::Status(StatusCode code, std::string msg, std::shared_ptr<StatusDetail> detail) {
   ARROW_CHECK_NE(code, StatusCode::OK) << "Cannot construct ok status with message";
-  state_ = new State;
-  state_->code = code;
-  state_->msg = std::move(msg);
-  if (detail != nullptr) {
-    state_->detail = std::move(detail);
-  }
+  state_ = new State{code, std::move(msg), std::move(detail)};
 }
 
+// We would prefer that this destructor *not* be inlined, since the vast majority of
+// Statuses are OK and so inlining would superflously increase binary size.
+Status::State::~State() = default;
```

Comment:
My guess is that it's better to have this inlined, but avoid inlining `DeleteState` (which is effectively what the previous implementation was achieving with

```cpp
    if (ARROW_PREDICT_FALSE(state_ != NULL)) {
      DeleteState();
    }
```

Reply:
I'll try writing out the conditions and keeping DeleteState non-inline